lambdas/backend-api/src/templates/domain/template/template-repository.ts

```diff
@@ -0,0 +1,203 @@
+    }
+
+    return success(response?.Item as Template);
+  } catch (error) {
+    return failure(ErrorCase.DATABASE_FAILURE, 'Failed to get template', error);
+  }
+};
+
+const create = async (
+  template: CreateTemplate,
+  owner: string
+): Promise<ApplicationResult<Template>> => {
+  const entity: Template = {
+    ...template,
+    id: uuidv4(),
+    owner,
+    version: 1,
+    templateStatus: TemplateStatus.NOT_YET_SUBMITTED,
+    createdAt: new Date().toISOString(),
+    updatedAt: new Date().toISOString(),
+  };
+
+  try {
+    const response = await client.send(
+      new PutCommand({
+        TableName: process.env.TEMPLATES_TABLE_NAME,
+        Item: entity,
+        ReturnConsumedCapacity: 'TOTAL',
+      })
+    );
+
+    const consumedUnits = response?.ConsumedCapacity?.CapacityUnits || 0;
+
+    // Note: This is a bit strange, PutCommand does not return the any data when creating
+    // a new item. So we can only infer the success by checking the CapacityUnits
+    // Or do a GetItem - which is expensive.
+    if (consumedUnits === 0) {
+      throw new Error('Expected DynamoDB CapacityUnits to be greater than 0', {
+        cause: response,
+      });
+    }
+
+    return success(entity);
+  } catch (error) {
+    return failure(
+      ErrorCase.DATABASE_FAILURE,
+      'Failed to create template',
+      error
+    );
+  }
+};
+
+const update = async (
+  templateId: string,
+  template: UpdateTemplate,
+  owner: string
+): Promise<ApplicationResult<Template>> => {
+  const updateExpression = [
+    '#name = :name',
+    '#message = :message',
+    '#updatedAt = :updateAt',
+    '#templateStatus = :templateStatus',
+  ];
+
+  let expressionAttributeNames: Record<string, string> = {
+    '#name': 'name',
+    '#message': 'message',
+    '#templateStatus': 'templateStatus',
+    '#updatedAt': 'updatedAt',
+    '#templateType': 'templateType',
+  };
+
+  let expressionAttributeValues: Record<string, string> = {
+    ':name': template.name,
+    ':message': template.message,
+    ':templateStatus': template.templateStatus,
+    ':updateAt': new Date().toISOString(),
+    ':not_yet_submitted': TemplateStatus.NOT_YET_SUBMITTED,
+    ':templateType': template.templateType,
+  };
+
+  if (template.subject) {
+    updateExpression.push('#subject = :subject');
+    expressionAttributeNames = {
+      ...expressionAttributeNames,
+      '#subject': 'subject',
+    };
+    expressionAttributeValues = {
+      ...expressionAttributeValues,
+      ':subject': template.subject,
+    };
+  }
+
+  const input: UpdateCommandInput = {
+    TableName: process.env.TEMPLATES_TABLE_NAME,
+    Key: {
+      id: templateId,
+      owner,
+    },
+    UpdateExpression: `SET ${updateExpression.join(', ')}`,
+    ExpressionAttributeNames: expressionAttributeNames,
+    ExpressionAttributeValues: expressionAttributeValues,
+    ConditionExpression:
+      'attribute_exists(id) AND #templateStatus = :not_yet_submitted AND #templateType = :templateType',
```

Comment:
Attribute exists is here because the `UpdateCommand` will also create a new record if it does not exist. Which is really counterintuitive!